### PR TITLE
T86 add support ncbi refsnp api

### PR DESCRIPTION
**Pull request by jooolia**
## Description
Adds in support for the NCBI variation services api (https://api.ncbi.nlm.nih.gov/variation/v0/) for use in retrieving information about refsnps. 

The parameters used in this PR are based on changes made, but that we never merged in PR#88 where @sinarueeger made many changes to ncbi_snp_query().  I will look at again at these changes, but I think we should merge them into dev and then go from there (we can continue that discussion on 88 but just wanted to note that this was very much based on that work). 

## Notes: 
1) I added some extra values returned by the function to be able to help assess the output and the variants that we get out. These values are seqname, hgvs, assembly, and ref_seq. I do not necessarily think these params should be kept in a final version, but perhaps helpful for understanding what we think the outputs should be. 

There are quite a few open questions related to this change:
1) Should we also include clinical significance as reported by Clinvar (accessible with the https://www.ncbi.nlm.nih.gov/snp/ interface)
2) I do not feel that the representations included here are all very consistent. Does what we report match with what is reported in https://www.ncbi.nlm.nih.gov/snp/?
3) Should we limit the number of queries that a user can send? 
4) Should we provide support for getting the hg19/37 position (if available)?
5) Should the new internal functions be exported? I do not see a reason to, but maybe they could be useful. So then maybe the "getting" of info from the api could be split into its own function as well. 

Ok this is a start. How can this be improved and be useful for the package?

## Related:
#88 

Thanks! Julia

### CODE REVIEW

**Comment by sinarueeger** [comment]
Hi @jooolia, 
This approach looks good to me! Let's use this as a base to finish up the changes. 

I added some comments to parts of the functions. 

- Regarding `seqname, hgvs, assembly, and ref_seq` I actually think that we could leave these, found them helpful so far. 
-  _report clinical significance_ yes
- _I do not feel that the representations included here are all very consistent. Does what we report match with what is reported in https://www.ncbi.nlm.nih.gov/snp/?_ can you make an example?
- _Should we limit the number of queries that a user can send?_ do you mean in one go? Maybe we could print a message saying that  passing on 100 SNPs will take > 100 seconds and we recommend to break the command up. 
- _Should we provide support for getting the hg19/37 position (if available)?_ yes!
- _Should the new internal functions be exported?_ I would not prioritize this for the moment. If people want more info, they might as well access the API themselves. 

What I'd like to additionally implement, is to access different allele frequencies. So instead of GnomAD by default, get the whole list of whats available. But thats not a priority either :-)

File: R/ncbi_snp_api.R [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+      warning(Query, " has been merged into ", Marker, call. = FALSE)
+      
+      variant.url <- paste0("https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/", no_rs_Marker)
+      variant.response <- httr::GET(variant.url)
+      variant.response.content <- RJSONIO::fromJSON(rawToChar(variant.response$content),
+                                                    simplifyWithNames = TRUE)
+      Class <- as.character(variant.response.content$primary_snapshot_data$variant_type)
+      placement_SNP <- get_placements(variant.response.content)
+      
+      if (Class %in% c("snv", "snp", "delins")) {
```
Comment (sinarueeger) [inline R/ncbi_snp_api.R:252]: Here are repeated `frequency_SNP <- ` elements. Have you thought about reducing this by including `if(Class %in%...) `into the `get_frequency` function?

File: R/ncbi_snp_api.R [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+      Marker <- as.character(paste0("rs", variant.response.content$merged_snapshot_data$merged_into))
+      no_rs_Marker <-  as.character(variant.response.content$merged_snapshot_data$merged_into)
+      
+      warning(Query, " has been merged into ", Marker, call. = FALSE)
+      
+      variant.url <- paste0("https://api.ncbi.nlm.nih.gov/variation/v0/refsnp/", no_rs_Marker)
+      variant.response <- httr::GET(variant.url)
+      variant.response.content <- RJSONIO::fromJSON(rawToChar(variant.response$content),
+                                                    simplifyWithNames = TRUE)
+      Class <- as.character(variant.response.content$primary_snapshot_data$variant_type)
```
Comment (sinarueeger) [inline R/ncbi_snp_api.R:249]: I think you could close the if condition here and jump directly to line 275. But maybe I am simplifying things.

File: man/ncbi_snp_query_api.Rd [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+\value{
+A dataframe with columns:
+\itemize{
+\item Query: The rs ID that was queried.
+\item Chromosome: The chromosome that the marker lies on.
+\item BP: The chromosomal position, in base pairs, of the marker,
+as aligned with the current genome used by dbSNP. we add 1 to the base
+pair position in the BP column in the output data.frame to agree with
+what the dbSNP website has.
+\item Marker: The name of the marker. If the rs ID queried
```
Comment (sinarueeger) [inline man/ncbi_snp_query_api.Rd:23]: Have you thought about renaming marker with rsid? or snp? I think "marker" is a bit outdated.

File: R/ncbi_snp_api.R (hunk 1 above)
Comment (jooolia) [reply R/ncbi_snp_api.R:252]: Very nice idea! makes things much more straightforward.

File: R/ncbi_snp_api.R (hunk 2 above)
Comment (jooolia) [reply R/ncbi_snp_api.R:249]: Brilliant. :) yes a lot of repeated code here that could be removed with this suggestion.

**Comment by jooolia** [comment]
> Hi @jooolia,
> This approach looks good to me! Let's use this as a base to finish up the changes.
>
Great! Ok I am keen to get your changes from #88 incorporated as well and to retire some of the failing functions. Babysteps. :)

 
> I added some comments to parts of the functions.

I have done my best to address your helpful comments. I will open up a few other issues so that we can work away at some more additions here. 

> 
> * Regarding `seqname, hgvs, assembly, and ref_seq` I actually think that we could leave these, found them helpful so far.
ok. 
> * _report clinical significance_ yes
ok I will add an issue
> * _I do not feel that the representations included here are all very consistent. Does what we report match with what is reported in https://www.ncbi.nlm.nih.gov/snp/?_ can you make an example?
I will create another issue to check this.
> * _Should we limit the number of queries that a user can send?_ do you mean in one go? Maybe we could print a message saying that  passing on 100 SNPs will take > 100 seconds and we recommend to break the command up.
> * _Should we provide support for getting the hg19/37 position (if available)?_ yes!
I think this should be not so difficult, but I will also create an issue
> * _Should the new internal functions be exported?_ I would not prioritize this for the moment. If people want more info, they might as well access the API themselves.
ok. 
> 
> What I'd like to additionally implement, is to access different allele frequencies. So instead of GnomAD by default, get the whole list of whats available. But thats not a priority either :-)
Yes nice idea!

Let me know if there are other parts of this that need attention. Thanks, Julia

**Comment by jooolia** [comment]
Oops just re-read this: "For parsing JSON, use jsonlite instead of rjson or RJSONIO." from ROpenSci. Will adapt tomorrow.

**Comment by jooolia** [comment]
Hi @sinarueeger ,
OK to convert between RJSONIO and jsonlite will require a few changes, but I can't get to them today. I would vote more for finishing the changes related to what is in this PR and I will open another issue for updating the code to support jsonlite. What do you think?

**Comment by sinarueeger** [comment]
I think thats a good idea. Let's use the working version and tackle the fine grained bits later.

**Comment by sinarueeger** [comment]
@jooolia I let you delete the branch if you don't need it anymore.

**Comment by jooolia** [comment]
ok great @sinarueeger ! I will delete the branch. I am of the opinion that merged branches can be deleted. I will see if I can clean up any other branches too.

**Pull request merged**